tests/robustness/attackers/test_classification_attackers.py

```diff
@@ -0,0 +1,103 @@
+from holisticai.robustness.metrics import adversarial_accuracy, empirical_robustness
+from sklearn.metrics import accuracy_score
+import pytest
+
+SHARD_SIZE=60
+
+@pytest.fixture
+def categorical_dataset():
+    dataset = load_dataset("adult", protected_attribute='race') # x,y,group_a,group_b
+    dataset = dataset.groupby(['y','group_a']).sample(SHARD_SIZE, random_state=42) # 0-ga | 0-gb  | 1-ga | 1-gb
+    train_test = dataset.train_test_split(test_size=0.02, stratify=dataset['y'], random_state=0)
+    train = train_test['train']
+    test = train_test['test']
+    correlations = train['X'].corrwith(train['y']).sort_values(ascending=False)
+    top_10_features = correlations.head(10).index.tolist()
+    train['X'] = train['X'][top_10_features]
+    test['X'] = test['X'][top_10_features]
+    return train['X'], test['X'], train['y'], test['y']
+
+
+def test_hsj(categorical_dataset):
+    train_X, test_X, train_y, test_y = categorical_dataset
+
+    # Standardize data and fit model
+    scaler = StandardScaler()
+    X_train = scaler.fit_transform(train_X)
+
+    X_test = scaler.transform(test_X)
+    feature_names = list(test_X.columns)
+
+    y_train = train_y
+    y_test = test_y
+
+    model = LogisticRegression()
+    model.fit(X_train, y_train)
+    y_pred = model.predict(X_test)
+    baseline_accuracy = accuracy_score(y_test, y_pred)
+
+    proxy = BinaryClassificationProxy(predict=model.predict, predict_proba=model.predict_proba, classes=[0, 1])
+
+    hsj_attacker = HopSkipJump(name="HSJ", predictor=proxy.predict)
+
+    hsj_adv_x = hsj_attacker.generate(pd.DataFrame(X_test, columns=feature_names))
+    y_adv_pred = proxy.predict(hsj_adv_x)
+
+    hsj_accuracy = adversarial_accuracy(y_test, y_pred, y_adv_pred)
+    hsj_robustness = empirical_robustness(X_test, hsj_adv_x, y_pred, y_adv_pred, norm=2)
+
+    # This test should pass when both dataframes are different
+    assert not hsj_adv_x.equals(pd.DataFrame(X_test, columns=feature_names))
+    # This test should pass when both accuracies are different
+    assert hsj_accuracy != baseline_accuracy
+    # This test should pass when the robustness is greater than 0
+    assert hsj_robustness > 0
+
+def test_zoo(categorical_dataset):
+    train_X, test_X, train_y, test_y = categorical_dataset
+
+    # Standardize data and fit model
+    scaler = StandardScaler()
+    X_train = scaler.fit_transform(train_X)
+
+    X_test = scaler.transform(test_X)
+    feature_names = list(test_X.columns)
+
+    y_train = train_y
+    y_test = test_y
+
+    model = LogisticRegression()
+    model.fit(X_train, y_train)
+    y_pred = model.predict(X_test)
+    baseline_accuracy = accuracy_score(y_test, y_pred)
+
+    proxy = BinaryClassificationProxy(predict=model.predict, predict_proba=model.predict_proba, classes=[0, 1])
+
+    predict_proba_fn = format_function_predict_proba(proxy.learning_task, proxy.predict_proba)  # type: ignore
```

Comment:
proxy object must be passed to ZooAttack. format_function_predict_proba must be an internal function

Reply:
updated